### ISSUE TITLE
ArrayIndexOutOfBoundsException thrown for MultiCurve with srsDimension=3

### ISSUE THREAD

**Issue by heidivanparys** [opengeospatial/ets-gml32#32]
A java.lang.ArrayIndexOutOfBoundsException is thrown in test valid Line String when testing this file: 
[multicurve_z.txt](https://github.com/opengeospatial/ets-gml32/files/2318252/multicurve_z.txt)

I think that it has something to do with the fact that z coordinates are given as well, I don't get this exception when testing a MultiCurve with only x and y coordinates.

![billede](https://user-images.githubusercontent.com/11427611/44583753-315e5980-a796-11e8-90da-4817f9d9c857.png)

![billede](https://user-images.githubusercontent.com/11427611/44583772-463aed00-a796-11e8-936d-3edfcafc65b7.png)

**Comment by dstenger** [comment]
Thank you for reporting this issue. We will do further investigation.

**Comment by keshav-nangare** [comment]
@dstenger 

This issue is not with the test but its related to one of the dependencies. The dependency geomatics-geotk is having the problem with following line:

https://github.com/opengeospatial/geomatics-geotk/blob/3e400ec87a46ab947221840a19ad5b1320598539/src/main/java/org/opengis/cite/geomatics/gml/GmlUtils.java#L255-L257

According to document the lenth of posListis 9 and dimension is 3 but this method has the crsDim=2 so it trying to process the 10th element from the values[i+1] array. So we need to fix this issue.

**Comment by lgoltz** [comment]
@keshav-nangare 
I agree with your analysis. I assume the failure is in https://github.com/opengeospatial/geomatics-geotk/blob/3e400ec87a46ab947221840a19ad5b1320598539/src/main/java/org/opengis/cite/geomatics/gml/CurveCoordinateListFactory.java#L134-L135:

The ccordinates are parsed from the posList, but the srsDimenstion is taken from the LineString (which defaults to 2 as it is not set. Correct should be (untested):

`GmlUtils.extractCoordinatesFromPosList(lineString.getPosList().getValue(), lineString.getPosList().getCoordinateDimension(), coords);`

**Comment by keshav-nangare** [comment]
@dstenger @lgoltz 

I have investigated this issue and found the problem with the dependency of geotoolkit.

The dependency geomatic-geotk is using the geotoolkit dependency to parse the gml resource and LineString class belongs to the geotoolkit/geotk-xml-gml dependency. 

As per the current requirement of the geomatic-geotk dependency it should return the srsDimentsion but this implementation is not found with the current version of the geotoolkit.

I have investigated the geotoolkit repository so they have added fix into master, I think the maven dependency is not released yet. Maybe we need to wait for the new release.

**Comment by lgoltz** [comment]
> I have investigated the geotoolkit repository so they have added fix into master, 

Link to the fix: https://github.com/Geomatys/geotoolkit/blob/f9e4d6c495330281f349633d52c1b1deaa7b39ad/modules/jaxb-xml-binding/geotk-xml-gml/src/main/java/org/geotoolkit/gml/xml/AbstractGeometry.java#L40

This can help us to detect if the the srsDimension is set for the geometry (Element LineString). But I think it also important, that with commit https://github.com/Geomatys/geotoolkit/commit/0b235420f1ca1ad2c915d0f3b9a0cf32bc9b1ab0#diff-f94dc86576986f5dffe1dfa5f159c5c6 it is possible to retrieve the srsDimension from the posList Element. Butt this also needs a new release.

@heidivanparys What is the priority for this issue?

**Comment by heidivanparys** [comment]
@lgoltz It's an important but not urgent issue for me. So I'll leave the prioritising to you.

**Comment by ghobona** [comment]
@keshav-nangare Please confirm whether this issue has been solved.

**Comment by keshav-nangare** [comment]
I have investigated the issue and found the problem in the `unmarshal` method. While unmarshalling data file to `LineStringType` class it is not assigning value to the `srsDimension`(so it becomes null) field. So line object passed for further execution as you can see in below snippet:
https://github.com/opengeospatial/ets-gml32/blob/ee05a587b538842976070c77dfe40d5533b2e5e6/src/main/java/org/opengis/cite/iso19136/data/spatial/LineStringTests.java#L131-L141
As the line object contains a null value for the `srsDimension` field because of that `geotoolkit` assigns default value as 2.  As you can see in below snippet while extracting coordinates from tupleList, it tries to access the 10th index element this becomes failure for ArrayIndexOutOfBoundsException.
https://github.com/opengeospatial/geomatics-geotk/blob/3e400ec87a46ab947221840a19ad5b1320598539/src/main/java/org/opengis/cite/geomatics/gml/GmlUtils.java#L255-L257

Proposed solution: 
As the unmarshal method is setting null value for `srsDimension`, we can set by retrieving it from the provided data file if it exists. We tried this solution and test executed successfully.
The following changes need to be done:
```
// Set srsDimension from data file if present.
NodeList posList = lineElem.getElementsByTagNameNS(GML32.NS_NAME, "posList");
if (posList.getLength() > 0) {
	Element pos = (Element) posList.item(0);
	if (null != pos && pos.getAttribute("srsDimension") != "") {
		line.setSrsDimension(Integer.valueOf(pos.getAttribute("srsDimension")));
	}                          
}
```

@dstenger  @ghobona  
Please let me know your thought on this.

**Comment by ghobona** [comment]
@keshav-nangare The proposed solution looks fine.

**Comment by keshav-nangare** [comment]
@dstenger @ghobona 

The proposed solution will fix the ArrayIndexOutOfBoundException but test will fail with the following assertion:
```
org.geotoolkit.gml.xml.v321.LineStringType<br>[@gml:id='id.c1a43714-6f81-445d-9c44-164add8b09cd.geom.0'] is not covered by valid area of CRS: BOX2D(441211.7826225794 6050281.295718309, 693780.8908102543 6404831.6351263095)
```

As the input data provided in first comment lies outside the valid area of CRS. If we change the input data values which are starting from 72 to 68 then the test gets passed.

**Comment by keshav-nangare** [comment]
Fixed with this PR #54.
